File: misc/coin_exchange.py
class Solution:
    def coinChange(self, coins, amount2):
        def backtrack(amount=0,count=0,start=0,path=[]):
            print(amount, path)
            if amount == 0:
                if self.mincoin > count:
                    self.res = path
                    self.mincoin = count
            elif amount > 0:
                for i in range(start,len(coins)):
                    if coins[i][1] > 0:
                        coins[i][1] -= 1
                        backtrack(amount - coins[i][0], count + 1, i, path + [coins[i][0]] )
                        coins[i][1] += 1
                        
                         
        self.mincoin = 1000
        self.res = []
        coins.sort(key=lambda x: x[0],reverse=True)
        backtrack(amount2)
        print(self.res)
        return -1 if self.mincoin == float("inf") else self.mincoin
    
s = Solution()
print(s.coinChange([[4,1],[2,6],[5,3]],22))

File: misc/coin_exchange.js
var s;

class Solution {
  coinChange(coins, amount2) {
    var backtrack;

    backtrack = (amount = 0, count = 0, start = 0, path = []) => {
      if (amount === 0) {
        if (this.mincoin > count) {
          this.res = path;
          this.mincoin = count;
        }
      } else {
        if (amount > 0) {
          for (var i = start, _pj_a = coins.length; i < _pj_a; i += 1) {
            if (coins[i][1] > 0) {
              coins[i][1] -= 1;
              backtrack(amount - coins[i][0], count + 1, i, path + [coins[i][0]]);
              coins[i][1] += 1;
            }
          }
        }
      }
    };

    this.mincoin = Number.parseFloat("inf");
    this.res = [];
    coins.sort();
    backtrack(amount2);
    console.log(this.res);
    return this.mincoin === Number.parseFloat("inf") ? -1 : this.mincoin;
  }

}

s = new Solution();
console.log(s.coinChange([[4, 1], [2, 6], [5, 3]], 22));
